Filter out virtual devices from the speaker list

youennf opened the pull request:
#### f2949ad3bb1cf358d66c07b773feecdc1b99526b
<pre>
Filter out virtual devices from the speaker list
<a href="https://rdar.apple.com/142577212">rdar://142577212</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=285670">https://bugs.webkit.org/show_bug.cgi?id=285670</a>

Reviewed by Andy Estes and Eric Carlson.

Some applications install virtual audio drivers and we should not expose them as speaker.
Like done for microphone, we use labels to filter out webex and MS Teams virtual devices.
Manually tested.

* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
(WebCore::isVirtualDeviceFromLabel):
(WebCore::isValidMicrophoneDevice):
(WebCore::isValidSpeakerDevice):
(WebCore::computeAudioDeviceList):
(WebCore::isValidCaptureDevice): Deleted.

Canonical link: <a href="https://commits.webkit.org/288790@main">https://commits.webkit.org/288790@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86e6c45d9423b8b8e25f36cd47b84ff3487dbd54

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84427 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38732 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89508 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35438 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12034 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65660 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23504 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87473 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3113 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76703 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/45954 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3063 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30934 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34485 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31704 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11694 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8513 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74111 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72527 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73312 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18142 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17645 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16089 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3111 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11646 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17122 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11495 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14971 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13268 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
```diff
@@ -102,7 +102,12 @@ static bool deviceHasOutputStreams(AudioObjectID deviceID)
-static bool isValidCaptureDevice(const CoreAudioCaptureDevice& device, bool filterTapEnabledDevices)
+static bool isVirtualDeviceFromLabel(const String& label)
+{
+    return label.contains("WebexMediaAudioDevice"_s) || label.contains("Microsoft Teams Audio"_s);
```
I have a few thoughts:
1. I wonder if this should go in BrowserEngineKit instead of WebKit since other browsers likely want to apply the same filtering.
2. Is there a way for AVAudioSession to do this filtering for us so that we don't have to maintain a hardcoded filter list? Since other parts of the system filter out these devices (e.g., Control Center), it seems like this logic already exists elsewhere.

youennf replied:
This is macOS only so does not apply to AVAudioSession/BEK.
macOS system apps like sound settings app are doing filtering.

This only seems to be feasible in GPU process where we are doing capture/audio rendering and not UIProcess where we are doing enumeration.

We could try to move enumeration to UIProcess but it is also unclear whether this filtering would be too harsh, Alexei mentioned OBS for instance.

eric-carlson commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
```diff
@@ -150,8 +155,19 @@ static bool isValidCaptureDevice(const CoreAudioCaptureDevice& device, bool filt
+static bool isValidSpeakerDevice(const CoreAudioCaptureDevice& device)
+{
+    if (isVirtualDeviceFromLabel(device.label())) {
+        RELEASE_LOG(WebRTC, "Ignoring virtual audio device");
```
Nit: `isVirtualDeviceFromLabel` will have already logged that the device is being ignored

youennf replied:
Not after the refactoring, but it might be better to make the logging specific to speaker. I'll fix that.

youennf committed:
Filter out virtual devices from the speaker list
rdar://142577212
https://bugs.webkit.org/show_bug.cgi?id=285670

Reviewed by Andy Estes and Eric Carlson.

Some applications install virtual audio drivers and we should not expose them as speaker.
Like done for microphone, we use labels to filter out webex and MS Teams virtual devices.
Manually tested.

* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
(WebCore::isVirtualDeviceFromLabel):
(WebCore::isValidMicrophoneDevice):
(WebCore::isValidSpeakerDevice):
(WebCore::computeAudioDeviceList):
(WebCore::isValidCaptureDevice): Deleted.

Canonical link: https://commits.webkit.org/288790@main
